tests/snippets/subclass_str.py

```diff
@@ -0,0 +1,22 @@
+from testutils import assertRaises
+
+x = "An interesting piece of text"
+assert x is str(x)
+
+class Stringy(str):
+    def __new__(cls, value=""):
+        return str.__new__(cls, value)
+
+    def __init__(self, value):
+        self.x = "substr"
+
+y = Stringy(1)
+assert type(y) is Stringy, "Type of Stringy should be stringy"
+assert type(str(y)) is str, "Str of a str-subtype should be a str."
+
+assert y + " other" == "1 other"
+assert y.x == "substr"
+
+## Base strings currently get an attribute dict, but shouldn't.
+# with assertRaises(AttributeError):
+#     "hello".x = 5
```

Comment:
I think it is more consistent that base strings have attribute dicts, I believe cpython is wrong here. We can deviate. You can use the platform module to check for rustpython.

Reply:
It feels like a pretty serious performance question, if we have to allocate dicts for intermediate ints and bools in computations.

Reply:
It's important enough that's ```__slots__``` exists so that user classes can take advantage of this.

Reply:
> It feels like a pretty serious performance question, if we have to allocate dicts for intermediate ints and bools in computations.

Not just that; this also makes other common optimizations on builtin types (string interning, small int cache) much weirder and harder to reason about.

```
>>> a = "hello"
>>> b = "hello"
>>> a is b
True
>>> a = 123
>>> b = 123
>>> a is b
True
```

So if in one section of code I do `"hello".x = 5`, this may affect any other code (including third-party library) using the same "hello" literal - as it may or may not be interned. Same with other builtins. That's implicit global state.